Remotion/ObjectBinding/UnitTests/Design/DropDownEditorBaseTest.cs

```diff
@@ -17,143 +17,172 @@
 
     [SetUp]
     public void SetUp ()
     {
-      _mockRepository = new MockRepository();
-
-      _mockTypeDescriptorContext = _mockRepository.StrictMock<ITypeDescriptorContext>();
-      _mockServiceProvider = _mockRepository.StrictMock<IServiceProvider>();
-      _mockWindowsFormsEditorService = _mockRepository.StrictMock<IWindowsFormsEditorService>();
-      _mockDropDownEditorBase = _mockRepository.PartialMock<MockDropDownEditorBase>();
-      _mockEditorControlBase = _mockRepository.PartialMock<EditorControlBase> (_mockServiceProvider, _mockWindowsFormsEditorService);
+      _mockTypeDescriptorContext = new Mock<ITypeDescriptorContext> (MockBehavior.Strict);
+      _mockServiceProvider = new Mock<IServiceProvider> (MockBehavior.Strict);
+      _mockWindowsFormsEditorService = new Mock<IWindowsFormsEditorService> (MockBehavior.Strict);
+      _mockDropDownEditorBase = new Mock<MockDropDownEditorBase>() { CallBase = true };
+      _mockEditorControlBase = new Mock<EditorControlBase> (_mockServiceProvider.Object, _mockWindowsFormsEditorService.Object) { CallBase = true };
     }
 
     [Test]
     public void EditValue ()
     {
-      Expect.Call (_mockTypeDescriptorContext.Instance).Return (new object());
-      Expect.Call (_mockServiceProvider.GetService (typeof (IWindowsFormsEditorService))).Return (_mockWindowsFormsEditorService);
-      Expect.Call (_mockDropDownEditorBase.NewCreateEditorControl (_mockTypeDescriptorContext, _mockWindowsFormsEditorService))
-          .Return (_mockEditorControlBase);
-      using (_mockRepository.Ordered())
-      {
-        _mockEditorControlBase.Value = "The input value";
-        _mockWindowsFormsEditorService.DropDownControl (_mockEditorControlBase);
-        Expect.Call (_mockEditorControlBase.Value).Return ("The output value");
-      }
-      _mockRepository.ReplayAll();
-
-      object actual = _mockDropDownEditorBase.EditValue (_mockTypeDescriptorContext, _mockServiceProvider, "The input value");
-
-      _mockRepository.VerifyAll();
+      _mockTypeDescriptorContext.Setup (_ => _.Instance).Returns (new object()).Verifiable();
+      _mockServiceProvider.Setup (_ => _.GetService (typeof (IWindowsFormsEditorService))).Returns (_mockWindowsFormsEditorService.Object).Verifiable();
+      _mockDropDownEditorBase.Setup (_ => _.NewCreateEditorControl (_mockTypeDescriptorContext.Object, _mockWindowsFormsEditorService.Object))
+          .Returns (_mockEditorControlBase.Object)
+          .Verifiable();
+      var sequence = new MockSequence();
+      _mockEditorControlBase.SetupProperty (_ => _.Value);
+      _mockWindowsFormsEditorService.InSequence (sequence)
+          .Setup (_ => _.DropDownControl (_mockEditorControlBase.Object))
+          .Callback (
+              (Control control) =>
+              {
+                Assert.That (_mockEditorControlBase.Object.Value.Equals ("The input value"));
+
+                _mockEditorControlBase.InSequence (sequence)
```

Comment:
Ah, okay, that works?

I was thinking we don't even need the entire setup incl. the sequence since we already check the Value.

Reply:
The setup is needed because `Value` should return `The output value` and not `The input value`.

The code looks like this, at first `control.Value` is set to `The input value` and then `value` gets changed to `control.Value` which returns `The output value` because of the `Setup` inside the `Callback`
Without the `Setup` the test would fail because the value of `value` would not be set to `The output value`.
```
control.Value = value;
editorService.DropDownControl (control);
value = control.Value;
...
return value;
```